### PR TITLE
Update tables for better use of API filters, context cancellation in list calls and page limiting for limit clause in query. Closes #336

### LINKED ISSUE #336: Update tables for better use of API filters, context cancellation in list calls and page limiting for limit clause in query.
Tables to be updated:

- gcp_bigquery_dataset
- gcp_bigquery_job
- gcp_bigquery_table
- gcp_cloudfunctions_function
- gcp_compute_address
- gcp_compute_backend_bucket
- gcp_compute_backend_service
- gcp_compute_disk
- gcp_compute_firewall
- gcp_compute_forwarding_rule
- gcp_compute_global_address
- gcp_compute_global_forwarding_rule
- gcp_compute_instance_template
- gcp_compute_machine_type
- gcp_compute_network
- gcp_compute_node_group
- gcp_compute_node_template
- gcp_compute_region
- gcp_compute_resource_policy
- gcp_compute_router
- gcp_compute_snapshot
- gcp_compute_ssl_policy
- gcp_compute_subnetwork
- gcp_compute_target_https_proxy
- gcp_compute_target_pool
- gcp_compute_target_ssl_proxy
- gcp_compute_target_vpn_gateway
- gcp_compute_url_map
- gcp_compute_vpn_tunnel
- gcp_compute_zone
- gcp_dns_managed_zone
- gcp_dns_policy
- gcp_dns_record_set
- gcp_kms_key_ring
- gcp_kms_key
- gcp_logging_bucket
- gcp_logging_exclusion
- gcp_logging_metric
- gcp_logging_sink
- gcp_monitoring_alert_policy
- gcp_monitoring_group
- gcp_monitoring_notification_channel
- gcp_organization
- gcp_project_organization_policy
- gcp_project_service
- gcp_pubsub_snapshot
- gcp_pubsub_subscription
- gcp_pubsub_topic
- gcp_service_account
- gcp_sql_backup
- gcp_sql_database_instance

### PR DESCRIPTION
**Pull request by c0d3r-arnab**
# Integration test logs
[integration_test.txt](https://github.com/turbot/steampipe-plugin-gcp/files/7861985/integration_test.txt)

# Example query results
[output.txt](https://github.com/turbot/steampipe-plugin-gcp/files/7861984/output.txt)

### CODE REVIEW

**Review by ParthaI — APPROVED** [review]
LGTM

File: gcp/table_gcp_cloudfunctions_function.go [hunk 1]
```diff
@@ -210,6 +210,18 @@ func listCloudFunctions(ctx context.Context, d *plugin.QueryData, h *plugin.Hydr
 		return nil, err
 	}
 
+	pageSize := types.Int64(1000)
```
Comment (rajlearner17) [inline gcp/table_gcp_cloudfunctions_function.go:213]: This `pageSize` is set to 1000, we should comment why we have set this

File: gcp/table_gcp_compute_address.go [hunk 2]
```diff
@@ -150,6 +157,33 @@ func listComputeAddresses(ctx context.Context, d *plugin.QueryData, h *plugin.Hy
...
+	}
+
+	filters := buildQueryFilterFromQuals(filterQuals, d.Quals)
+	filterString := ""
+	if len(filters) > 0 {
+		filterString = strings.Join(filters, " ")
+	}
+	plugin.Logger(ctx).Trace("listComputeAddresses", "filterString", filterString)
+
+	pageSize := types.Int64(500)
```
Comment (rajlearner17) [inline gcp/table_gcp_compute_address.go:175]: pageSize := types.Int64(500) can you provide the doc link in the comment

File: gcp/table_gcp_compute_instance.go [hunk 3]
```diff
@@ -312,7 +312,11 @@ func listComputeInstances(ctx context.Context, d *plugin.QueryData, h *plugin.Hy
 	limit := d.QueryContext.Limit
 	if d.QueryContext.Limit != nil {
 		if *limit < *pageSize {
-			pageSize = limit
```
Comment (rajlearner17) [inline gcp/table_gcp_compute_instance.go:315]: Remove the line no 309  **plugin.Logger(ctx).Trace("listComputeInstances", "IAM I HERE")**

File: gcp/table_gcp_compute_global_address.go [hunk 4]
```diff
@@ -147,6 +155,32 @@ func listComputeGlobalAddresses(ctx context.Context, d *plugin.QueryData, h *plu
...
+	if len(filters) > 0 {
+		filterString = strings.Join(filters, " ")
+	}
+	plugin.Logger(ctx).Trace("listComputeGlobalAddresses", "filter string", filterString)
+
+	pageSize := types.Int64(500)
+	limit := d.QueryContext.Limit
+	if d.QueryContext.Limit != nil {
+		if *limit < *pageSize {
+			if *limit < 1 {
```
Comment (rajlearner17) [inline gcp/table_gcp_compute_global_address.go:176]: Revisit -- do we need this in GCP?

File: gcp/table_gcp_compute_global_forwarding_rule.go [hunk 5]
```diff
@@ -189,6 +202,35 @@ func listComputeGlobalForwardingRules(ctx context.Context, d *plugin.QueryData,
...
+	}
+	plugin.Logger(ctx).Trace("listComputeGlobalForwardingRules", "filter string", filterString)
+
+	pageSize := types.Int64(500)
+	limit := d.QueryContext.Limit
+	if d.QueryContext.Limit != nil {
+		if *limit < *pageSize {
+			if *limit < 1 {
+				pageSize = types.Int64(1)
+			} else {
```
Comment (rajlearner17) [inline gcp/table_gcp_compute_global_forwarding_rule.go:228]:
```if *limit < 1 {
				pageSize = types.Int64(1)
			} else ```

Revisit as GCP does not have any limit issue, we have not experienced this.

**Review by rajlearner17 — APPROVED** [review]
LGTM as per latest testing of all compliance & example queries in this branch

File: gcp/table_gcp_compute_backend_bucket.go [hunk 6]
```diff
@@ -121,6 +129,28 @@ func listComputeBackendBuckets(ctx context.Context, d *plugin.QueryData, h *plug
...
+		{"bucket_name", "bucketName", "string"},
+		{"enable_cdn", "enableCdn", "boolean"},
+	}
+
+	filters := buildQueryFilterFromQuals(filterQuals, d.Quals)
+	filterString := ""
+	if len(filters) > 0 {
+		filterString = strings.Join(filters, " ")
+	}
+	plugin.Logger(ctx).Trace("listComputeBackendBuckets", "filter string", filterString)
```
Comment (bigdatasourav) [inline gcp/table_gcp_compute_backend_bucket.go:142]:
```suggestion
```

File: gcp/table_gcp_compute_backend_service.go [hunk 7]
```diff
@@ -239,6 +248,30 @@ func listComputeBackendServices(ctx context.Context, d *plugin.QueryData, h *plu
...
+		{"session_affinity", "sessionAffinity", "string"},
+		{"enable_cdn", "enableCdn", "boolean"},
+	}
+
+	filters := buildQueryFilterFromQuals(filterQuals, d.Quals)
+	filterString := ""
+	if len(filters) > 0 {
+		filterString = strings.Join(filters, " ")
+	}
+	plugin.Logger(ctx).Trace("listComputeBackendServices", "filter string", filterString)
```
Comment (bigdatasourav) [inline gcp/table_gcp_compute_backend_service.go:263]:
```suggestion
```

File: gcp/table_gcp_compute_firewall.go [hunk 8]
```diff
@@ -176,6 +184,28 @@ func listComputeFirewalls(ctx context.Context, d *plugin.QueryData, h *plugin.Hy
...
+		{"direction", "direction", "string"},
+		{"disabled", "disabled", "boolean"},
+	}
+
+	filters := buildQueryFilterFromQuals(filterQuals, d.Quals)
+	filterString := ""
+	if len(filters) > 0 {
+		filterString = strings.Join(filters, " ")
+	}
+	plugin.Logger(ctx).Trace("listComputeFirewalls", "filter string", filterString)
```
Comment (bigdatasourav) [inline gcp/table_gcp_compute_firewall.go:197]:
```suggestion
```

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Pull request merged**